File: src/venvi/services/providers/odh.py
from datetime import datetime
from typing import Any

import httpx

from venvi.models.event import Event
from venvi.services.providers.base import BaseEventProvider


class ODHProvider(BaseEventProvider):
    @property
    def source_name(self) -> str:
        return "odh"

    async def fetch_events(self) -> list[dict[str, Any]]:
        api_url = "https://tourism.opendatahub.com/v1/Event"
        params = {"pagenumber": 1, "pagesize": 20}
        async with httpx.AsyncClient() as client:
            response = await client.get(api_url, params=params)
            response.raise_for_status()
            return response.json().get("Items", [])

    def map_event(self, raw: dict[str, Any]) -> Event:
        details = raw.get("Detail", {})

        def get_localized(obj: dict, key: str) -> str | None:
            for lang in ["en", "it", "de"]:
                val = obj.get(lang, {}).get(key)
                if val:
                    return val
            return None

        title = get_localized(details, "Title") or "Untitled Event"
        description = get_localized(details, "BaseText") or get_localized(
            details, "IntroText"
        )

        city = raw.get("ContactInfos", {}).get("en", {}).get("City") or "Unknown"

        image_url = None
        gallery = raw.get("ImageGallery", [])
        if gallery:
            image_url = gallery[0].get("ImageUrl")

        raw_id = raw.get("Id") or str(hash(title + str(raw.get("DateBegin"))))

        return Event(
            id=f"{self.source_name}:{raw_id}",
            title=title,
            description=description,
            date_start=datetime.fromisoformat(raw["DateBegin"])
            if raw.get("DateBegin")
            else datetime.now(),
            date_end=datetime.fromisoformat(raw["DateEnd"])
            if raw.get("DateEnd")
            else datetime.now(),
            location=city,
            url=f"https://opendatahub.com/events/{raw_id}",  # Placeholder URL
            image_url=image_url,
            source_name=self.source_name,
            source_id=raw_id,
            topics=[],
            category="general",
            is_new=True,
        )

File: providers/odh.go
// Package providers implements event data source providers for Venvi.
package providers

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// ODHProvider fetches events from the Open Data Hub tourism API.
// The Open Data Hub (opendatahub.com) provides tourism events from South Tyrol.
type ODHProvider struct {
	// BaseURL allows overriding the API endpoint for testing.
	BaseURL string
	// Client is the HTTP client used for requests.
	Client *http.Client
}

// NewODHProvider creates a new ODH provider with default settings.
func NewODHProvider() *ODHProvider {
	return &ODHProvider{
		BaseURL: "https://tourism.opendatahub.com/v1/Event",
		Client:  &http.Client{Timeout: 30 * time.Second},
	}
}

// SourceName returns "odh" as the identifier for Open Data Hub events.
func (p *ODHProvider) SourceName() string {
	return "odh"
}

// ODHResponse represents the API response structure from Open Data Hub.
type ODHResponse struct {
	TotalResults int              `json:"TotalResults"`
	Items        []map[string]any `json:"Items"`
}

// FetchEvents retrieves events from the Open Data Hub API.
func (p *ODHProvider) FetchEvents(ctx context.Context) ([]RawEvent, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, p.BaseURL, nil)
	if err != nil {
		return nil, fmt.Errorf("creating request: %w", err)
	}

	q := req.URL.Query()
	q.Set("pagenumber", "1")
	q.Set("pagesize", "50") // Increased page size
	q.Set("active", "true")
	q.Set("odalactive", "true")
	q.Set("datefrom", time.Now().Format("2006-01-02"))
	req.URL.RawQuery = q.Encode()

	resp, err := p.Client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("fetching events: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status: %d", resp.StatusCode)
	}

	var result ODHResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decoding response: %w", err)
	}

	events := make([]RawEvent, len(result.Items))
	for i, item := range result.Items {
		events[i] = RawEvent(item)
	}
	return events, nil
}

// MapEvent transforms raw ODH event data into a unified Event structure.
func (p *ODHProvider) MapEvent(raw RawEvent) *Event {
	// ODH events might not have a city in ContactInfos, so we fallback to "Unknown"
	// and let the helper try to find it.
	event := buildEventFromRaw(raw, p.SourceName(), "Unknown", "")

	// ODH-specific overrides if needed (currently none as buildEventFromRaw covers it)
	// But we need to ensure URL is correct if ID is present
	// ODH-specific overrides
	if event.ID != "" && event.URL == "" {
		event.URL = "https://opendatahub.com/events/" + event.ID
	}

	// Filter out bad quality events
	// 1. Title looks like a UUID (32 chars, hex)
	if len(event.Title) == 32 && isHex(event.Title) {
		return nil
	}
	// 2. No description (or very short)
	if len(event.Description) < 10 {
		return nil
	}

	return event
}

// isHex checks if a string is hexadecimal.
func isHex(s string) bool {
	for _, c := range s {
		if (c < '0' || c > '9') && (c < 'a' || c > 'f') && (c < 'A' || c > 'F') {
			return false
		}
	}
	return true
}
